fix: do not multicall allowances globally

shoom3301 opened the pull request:
# Summary

Fixes https://www.notion.so/cownation/App-requires-unnecessary-permit-signature-2318da5f04ca80f4965ec3af8c0034e8?source=copy_link

The problem turned out to be deeper than I expected.
Essentially, `BalancesAndAllowancesUpdater` was using SWR config for both balances and allowances multicalls. But this SWR config is only based on balances cache state, hence it should not be used for allowances.
Then I started checking, do we really need allowances multicall for all tokens globally?
It turned out - no.
We only need it in orders table (limit/twap) and tokens page.
In all other places we only need allowance for current sell token.

To fix that, I did:
 - added `useTokenAllowance()`  and used it instead of `useTokensAllowances()`
 - moved `useTokenAllowances()` to ordersTable and used it only there and in tokens page

Now `lib/balances-and-allowances` should actually be `lib/balances`, but I decided to not rename it in this PR to avoid distracting changes. We can do this refactoring later.

With that we get:
 - less RPC nodes load, because we don't make multicalls for allowances globaly (only in orders table and tokens page)
 - fixed initial case, now allowances updates every 10 seconds and just after approval tx mining

### Please review this PR commit-by-commit, it will be more clear what I did.

# To Test

1. Initial case: https://www.notion.so/cownation/App-requires-unnecessary-permit-signature-2318da5f04ca80f4965ec3af8c0034e8?source=copy_link
2. Unfillable (not enough allowance) warning in orders table
3. Approval state in tokens page
4. Please also check approval flow and permit signing


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Introduced new hooks for fetching and checking token allowances and balances, providing more accurate and up-to-date allowance information.
  * Added utilities to determine if a token amount has sufficient allowance for transactions.

* **Refactor**
  * Simplified and centralized allowance retrieval and approval state logic for improved reliability and maintainability.
  * Streamlined trade and order flows to use the new allowance hooks, reducing complexity and dependencies.
  * Updated various components and hooks to use the new allowance and balance interfaces.

* **Bug Fixes**
  * Improved accuracy and consistency in allowance checks and approval state calculations across the app.
  * Enhanced safety in accessing allowance data to prevent potential runtime errors.

* **Chores**
  * Removed deprecated and redundant code related to legacy allowance management and balance checks.
  * Updated type imports and exports for better code organization and clarity.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

elena-zh reviewed (changes requested):
Hey @shoom3301 , the original issue is no longer reproducible! However, there are new ones: 
1. Tokens page: when no allowance is set, the app shows 0 approval amount+Approve all button
<img width="1780" height="830" alt="account tokens" src="https://github.com/user-attachments/assets/825397e8-c4db-4b67-8565-f0c135ca3ab8" />
 It should show 'Approve' button and should not show 0 amount 
2. I've placed a limit order with already approved permittable token: right after the order placement it says 'unfillable' (it is changed in 30 seconds)
<img width="1811" height="257" alt="allowance" src="https://github.com/user-attachments/assets/06f76c4f-aebd-48ac-8df6-d3fc452d8154" />
<img width="1416" height="133" alt="approved1" src="https://github.com/user-attachments/assets/11842cee-2368-4c4e-8046-daf9dad94ece" />
3. Incorrect token balance in forms: when I pick a token, the balance blinks with the previous tokent's balance. E.g. token's balance is >4, the form shows '0.005'. Due to this reason an order may show 'insufficient balance' on the Your orders table
<img width="540" height="222" alt="dai bal" src="https://github.com/user-attachments/assets/d74b76a8-93ae-4b07-a4a2-ed8df0d53470" />
<img width="633" height="551" alt="Screenshot_4" src="https://github.com/user-attachments/assets/533236bb-ecba-41ed-bcdc-1e98ac396283" />
<img width="953" height="241" alt="image" src="https://github.com/user-attachments/assets/b3486e79-4d64-42ca-a909-4bc6525a242a" />

4. If a permit is already signed, for example, on limit orders page, the permit hook is not displayed on the Swap confirm modal: 
<img width="794" height="639" alt="Screenshot_5" src="https://github.com/user-attachments/assets/afb0f9f2-828b-4d29-98df-0cc3fbc062cf" />
It should be displayed: see it on Dev: 
<img width="784" height="661" alt="this hook" src="https://github.com/user-attachments/assets/38ab6879-36c0-41e4-93c6-8bfec238ac4e" />
5. Also, hook details are missing on the Swap confirm modal when need to sign this request in the wallet: 
<img width="1195" height="573" alt="hook details" src="https://github.com/user-attachments/assets/bf85858f-15fc-493b-84eb-aa39cfe203ea" />
6. The app makes me sign the permit once again after successful swap TX with this token. E.g.
 - I placed swap order with a permittable token, there was permit hook inside
 - swap order got filled
 - I went to Limit orders and started placing an order with the same token --> 
<img width="1776" height="821" alt="new hook when approved" src="https://github.com/user-attachments/assets/c045a355-8190-4307-8f63-ac79a3591682" />
Reproducible only in 1st 30 seconds after the swap execution

7. Safe: the order is Unfillable after placement when I partially approved sell token (e.f. 50% of it). Order is partially fillable. Might be related to the case 2 above
<img width="1426" height="239" alt="image" src="https://github.com/user-attachments/assets/2a78945c-4312-4c85-bde4-8c322e6438e6" />

8. TWAP orders show 'Insufficient allowance' right after order placement , however, there is enough allowance
<img width="1308" height="281" alt="image" src="https://github.com/user-attachments/assets/7e16899d-a151-43bc-9819-eac258900ff8" />

Could you please take a look at these issues?

shoom3301 commented:
@elena-zh 

1. Fixed
2. Couldn't reproduce
3. Fixed
4,5,6: Cannot reproduce (see demo1 bellow)
7,8: Cannot reproduce
9: Cannot reproduce (see demo2 bellow)

Demo1:

https://github.com/user-attachments/assets/5489d472-3467-4f46-a696-969dac1b0a5e

Demo2:

https://github.com/user-attachments/assets/b7598d50-2fea-41d2-9a97-74e2032481ab

elena-zh reviewed (approved):
Thank you, looks good now

shoom3301 commented on apps/cowswap-frontend/src/legacy/components/Tokens/TokensTableRow.tsx:
```diff
@@ -147,7 +150,7 @@ export const TokensTableRow = ({
 
     if (!account || approvalState === ApprovalState.NOT_APPROVED) {
-      if (!currentAllowance || currentAllowance.equalTo(0)) {
+      if (!currentAllowanceAmount) {
```
This was already fixed in the following commits

shoom3301 commented on apps/cowswap-frontend/src/modules/tokens/hooks/useEnoughBalance.ts:
```diff
@@ -1,108 +1,54 @@
-  if (!tokenAddress || !allowances) {
-    return undefined
-  }
-  if (isNativeCurrency) {
-    return true
-  }
```
Thanks, added a case:

```
if (amount && getIsNativeToken(amount.currency)) {
      return true
    }
```
